### PR TITLE
ci: trigger sanity suite on pre post release

### PR DESCRIPTION
**Pull request by MoumitaM**
## PR Description

With this change after staging/production deployment sanity suite will run automatically on browserstack.

## Linear task (optional)

[Linear task link](https://linear.app/rudderstack/issue/SDK-2195/automate-js-sdk-sanity-test-on-browserstack)

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [ ] Chrome
- [ ] Firefox
- [ ] IE11

## Sanity Suite

- [ ] All sanity suite test cases pass locally

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new job to dynamically extract the SDK version in both production and staging deployment workflows.
	- Added a job to trigger BrowserStack sanity suite tests using the determined SDK version.
	- Created a new workflow for running sanity tests with user-defined environment and SDK version inputs.

- **Improvements**
	- Enhanced deployment workflows to adapt to different versioning scenarios, ensuring accurate SDK version testing.
	- Streamlined the testing workflow by allowing flexible environment selection and monorepo version management.

### CODE REVIEW

File: .github/workflows/deploy-staging.yml [hunk 1]
```diff
@@ -48,3 +48,41 @@ jobs:
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout source branch
+        uses: actions/checkout@v4
+
+      - name: Determine the SDK version
+        id: determine_version
+        run: |
+          # Default new_version to empty
```
Comment (saikumarrs) [inline .github/workflows/deploy-staging.yml:62]: Similar to the above comment, there is a neat way to extract the version easily. Let's discuss this offline.
For reference, see `deploy.yml` source code checkout steps. We just need to checkout the appropriate commit and read `package.json`. (ignore the fact that `use_pr_head_sha` input is not used for now)

Comment (saikumarrs) [reply .github/workflows/deploy-staging.yml:62]: In effect, this version calculation can be moved into the sanity workflow.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy-prod.yml [hunk 2]
```diff
@@ -43,3 +43,41 @@ jobs:
...
+            new_version="${new_version#release/}"  # Strip release/
+
+          # Check if it's a hotfix release branch
+          elif [[ "${{ github.head_ref }}" == hotfix-release/* ]]; then
+            new_version="${{ github.head_ref }}" 
+            new_version="${new_version#hotfix-release/}"  # Strip hotfix-release/
+          fi
+
+          # Set the version as an output variable
+          echo "SDK_VERSION=$new_version"
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/deploy-prod.yml]: Addressed

File: .github/workflows/deploy-prod.yml [hunk 3]
```diff
@@ -43,3 +43,41 @@ jobs:
...
+
+  trigger-bs-sanity-suite:
+    uses: ./.github/workflows/browserstack-sanity.yml
+    with:
+      environment: production
+      sdkVersion: $SDK_VERSION
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/deploy-prod.yml]: Addressed

File: .github/workflows/deploy-prod.yml [hunk 4]
```diff
@@ -43,3 +43,41 @@ jobs:
...
+          # Check if it's a hotfix release branch
+          elif [[ "${{ github.head_ref }}" == hotfix-release/* ]]; then
+            new_version="${{ github.head_ref }}" 
+            new_version="${new_version#hotfix-release/}"  # Strip hotfix-release/
+          fi
+
+          # Set the version as an output variable
+          echo "SDK_VERSION=$new_version"
+
+  trigger-bs-sanity-suite:
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/deploy-prod.yml]: Addressed

File: .github/workflows/deploy-prod.yml [hunk 5]
```diff
@@ -43,3 +43,41 @@ jobs:
       RS_PROD_BUGSNAG_API_KEY: ${{ secrets.RS_PROD_BUGSNAG_API_KEY }}
       SLACK_BOT_TOKEN: ${{ secrets.SLACK_BOT_TOKEN }}
       SLACK_RELEASE_CHANNEL_ID: ${{ secrets.SLACK_RELEASE_CHANNEL_ID_NON_PROD }}
+
+  extract-version:
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/deploy-prod.yml]: Addressed

File: .github/workflows/deploy-prod.yml [hunk 6]
```diff
@@ -43,3 +43,41 @@ jobs:
...
+    runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/deploy-prod.yml]: Addressed

File: .github/workflows/deploy-staging.yml [hunk 7]
```diff
@@ -48,3 +48,41 @@ jobs:
...
+          elif [[ "${{ github.head_ref }}" == hotfix-release/* ]]; then
+            new_version="${{ github.head_ref }}" 
+            new_version="${new_version#hotfix-release/}"  # Strip hotfix-release/
+          fi
+
+          # Set the version as an output variable
+          echo "SDK_VERSION=$new_version"
+
+  trigger-bs-sanity-suite:
+    uses: ./.github/workflows/browserstack-sanity.yml
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/deploy-staging.yml]: Addressed

File: .github/workflows/deploy-staging.yml (hunk 1 above)
Comment (MoumitaM) [reply .github/workflows/deploy-staging.yml]: Addressed

File: .github/workflows/deploy-staging.yml [hunk 8]
```diff
@@ -48,3 +48,41 @@ jobs:
...
+  extract-version:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout source branch
+        uses: actions/checkout@v4
+
+      - name: Determine the SDK version
+        id: determine_version
+        run: |
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/deploy-staging.yml]: Updated implementation.

File: .github/workflows/trigger-sanity-suite.yml [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      - name: Determine the SDK version
+        id: determine_version
+        run: |
+          if [ -n "${{ inputs.sdkVersion }}" ]; then
+            current_version="${{ inputs.sdkVersion }}"
+          else
+            current_version=$(jq -r .version package.json)
+          fi
+          echo "::set-output name=current_version::$current_version"
+
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/trigger-sanity-suite.yml]: Addressed

File: .github/workflows/trigger-sanity-suite.yml [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        with:
+          ref: ${{ steps.getSHA.outputs.SHA }}
+
+      - name: Determine the SDK version
+        id: determine_version
+        run: |
+          if [ -n "${{ inputs.sdkVersion }}" ]; then
+            current_version="${{ inputs.sdkVersion }}"
+          else
+            current_version=$(jq -r .version package.json)
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/trigger-sanity-suite.yml:53]: Addressed

File: .github/workflows/trigger-sanity-suite.yml [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+          fi
+          echo "::set-output name=current_version::$current_version"
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/trigger-sanity-suite.yml]: Addressed

File: .github/workflows/trigger-sanity-suite.yml [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+      sdkVersion:
+        description: Version of the SDK
+        required: true
+  workflow_call:
+    inputs:
+      environment:
+        type: string
+        required: true
+        default: production
+
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/trigger-sanity-suite.yml]: not required

File: .github/workflows/trigger-sanity-suite.yml [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+      - name: Determine the SDK version
+        id: determine_version
+        run: |
+          if [ -n "${{ inputs.monorepoVersion }}" ]; then
+            current_version="${{ inputs.monorepoVersion }}"
+          else
+            current_version=$(jq -r .version package.json)
+          fi
+          echo "current_version=$current_version" >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/trigger-sanity-suite.yml]: Not reqired

File: .github/workflows/trigger-sanity-suite.yml [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+      monorepoVersion:
+        description: Version of the SDK
+        required: true
+  workflow_call:
+    inputs:
+      environment:
+        type: string
+        required: true
+        default: production
+
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/trigger-sanity-suite.yml]: Not required

File: .github/workflows/trigger-sanity-suite.yml [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    needs: extract-monorepo-version
+
+    steps:
+      - name: Trigger sanity suite test workflow on rudder-client-side-test
+        run: |
+          curl -X POST \
+          -H "Authorization: token ${{ secrets.PAT }}" \
+          -H "Accept: application/vnd.github.v3+json" \
+          https://api.github.com/repos/rudderlabs/rudder-client-side-test/dispatches \
+          -d '{"event_type":"triggered_by_source_repo", "client_payload": {"environment": "${{ github.event.inputs.environment }}", "monorepoVersion": "${{ needs.extract-monorepo-version.outputs.version }}"}}'
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/trigger-sanity-suite.yml:67]: Incorrect

File: .github/workflows/trigger-sanity-suite.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (MoumitaM) [reply .github/workflows/trigger-sanity-suite.yml]: Incorrect

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**